scene/resources/material.cpp

```diff
@@ -954,132 +1000,200 @@ void BaseMaterial3D::_update_shader() {
 		code += "varying vec3 uv2_triplanar_pos;\n";
 	}
 	if (flags[FLAG_UV1_USE_TRIPLANAR]) {
-		code += "uniform float uv1_blend_sharpness;\n";
-		code += "varying vec3 uv1_power_normal;\n";
+		code += R"(
+uniform float uv1_blend_sharpness : hint_range(0.0, 150.0, 0.001);
+varying vec3 uv1_power_normal;
+)";
 	}
 
 	if (flags[FLAG_UV2_USE_TRIPLANAR]) {
-		code += "uniform float uv2_blend_sharpness;\n";
-		code += "varying vec3 uv2_power_normal;\n";
+		code += R"(uniform float uv2_blend_sharpness : hint_range(0.0, 150.0, 0.001);
+varying vec3 uv2_power_normal;
+)";
 	}
 
-	code += "uniform vec3 uv1_scale;\n";
-	code += "uniform vec3 uv1_offset;\n";
-	code += "uniform vec3 uv2_scale;\n";
-	code += "uniform vec3 uv2_offset;\n";
-
-	code += "\n\n";
+	code += R"(
+uniform vec3 uv1_scale;
+uniform vec3 uv1_offset;
+uniform vec3 uv2_scale;
+uniform vec3 uv2_offset;
 
-	code += "void vertex() {\n";
+void vertex() {
+)";
 
 	if (flags[FLAG_SRGB_VERTEX_COLOR]) {
-		code += "	if (!OUTPUT_IS_SRGB) {\n";
-		code += "		COLOR.rgb = mix(pow((COLOR.rgb + vec3(0.055)) * (1.0 / (1.0 + 0.055)), vec3(2.4)), COLOR.rgb * (1.0 / 12.92), lessThan(COLOR.rgb, vec3(0.04045)));\n";
-		code += "	}\n";
+		code += R"(	// Vertex Color is sRGB: Enabled
+	if (!OUTPUT_IS_SRGB) {
+		COLOR.rgb = mix(
+				pow((COLOR.rgb + vec3(0.055)) * (1.0 / (1.0 + 0.055)), vec3(2.4)),
+				COLOR.rgb * (1.0 / 12.92),
+				lessThan(COLOR.rgb, vec3(0.04045))
+		);
+	}
+)";
 	}
 	if (flags[FLAG_USE_POINT_SIZE]) {
-		code += "	POINT_SIZE=point_size;\n";
+		code += "\n	POINT_SIZE = point_size;\n";
 	}
 
 	if (shading_mode == SHADING_MODE_PER_VERTEX) {
-		code += "	ROUGHNESS=roughness;\n";
+		code += "	ROUGHNESS = roughness;\n";
 	}
 
 	if (!flags[FLAG_UV1_USE_TRIPLANAR]) {
-		code += "	UV=UV*uv1_scale.xy+uv1_offset.xy;\n";
+		code += "	UV = UV * uv1_scale.xy + uv1_offset.xy;\n";
 	}
```

Comment:
I'm wondering if we should systematically use R-strings (with a line break after `R"(` and before `");`) even for single line additions like this. This would add two extra lines of code each time, but the advantage would be that all generated code would start at column 0 in the file, and have its indentation convey the actual generated code indentation. This should make it easier to ensure that indentation is consistent.

Reply:
> I'm wondering if we should systematically use R-strings (with a line break after R"( and before ");) even for single line additions like this. 

If we do this, then every line of code will be separated by two blank lines in the generated shader since there are two newlines embedded in the string:
 
```
R("
something
")
```

We could do this to avoid always adding 2 newlines where relevant, but then the newline will be at the beginning of the string instead of the end:

```
R("
something")
```

Reply:
Hm yeah that's subpar either way. Amazing that even with a new feature like R-strings they couldn't figure out an actually user-friendly syntax :D

Reply:
To be fair, most languages with multiline strings have this issue. You typically need to call a method to trim leading/trailing whitespace if you don't want these to be part of the final string (which means it's not 100% compile-time anymore).